async_fuse/src/memfs/cache/mod.rs

```diff
@@ -0,0 +1,750 @@
+                if is_first_block {
+                    is_first_block = false;
+                    let off = offset.overflow_rem(self.block_size);
+                    block.overwrite_offset(
+                        off,
+                        (*buf).get(have_read..end).unwrap_or_else(|| {
+                            panic!("should not reach here, buf out of range in cache write")
+                        }),
+                    );
+                    have_read = end;
+                } else {
+                    block.overwrite((*buf).get(have_read..end).unwrap_or_else(|| {
+                        panic!("should not reach here, buf out of range in cache write")
+                    }));
+                    have_read = end;
+                }
+            }
+        };
+
+        for i in start_key..=end_key {
+            let s = if i == start_key {
+                bucket_start_offset
+            } else {
+                0
+            };
+            let l = if i == end_key {
+                bucket_end_offset.overflow_add(1)
+            } else {
+                self.bucket_size_in_block
+            };
+
+            if let Some(bucket) = file_cache.get(&i, &guard) {
+                self.queue
+                    .lock()
+                    .push(bucket.clone(), 0.overflow_sub(now_monotonic()));
+                bucket
+                    .write()
+                    .get_mut(s..l)
+                    .unwrap_or_else(|| {
+                        panic!("Just insert MemBlockBucket into file cache, but cannot get it.")
+                    })
+                    .iter_mut()
+                    .for_each(&mut copy_fn);
+            } else {
+                // Here might be a racing case
+                debug!(
+                    "memory_block_bucket for file {:?} index {} is empty, create one",
+                    OsStr::from_bytes(file_name),
+                    start_key
+                );
+                debug!("start offset {}, end offset {}", s, l);
+                let bucket = MemBlockBucket::new();
+                self.queue
+                    .lock()
+                    .push(bucket.clone(), 0.overflow_sub(now_monotonic()));
+                file_cache.insert(i, bucket);
+                file_cache
+                    .get(&i, &guard)
+                    .unwrap_or_else(|| {
+                        panic!("Just insert MemBlockBucket into file cache, but cannot get it.")
+                    })
+                    .write()
+                    .get_mut(s..l)
+                    .unwrap_or_else(|| panic!("should not reach here, out of range in cache write"))
+                    .iter_mut()
+                    .for_each(&mut copy_fn);
+            };
+        }
+
+        self.size
+            .fetch_add(grow_memory.overflow_mul(self.block_size), Ordering::Relaxed);
+
+        let mut dealloc_fn = |b: &mut Option<MemBlock>| {
+            if b.is_some() {
+                b.take();
```

Comment:
Maybe a dumb question, right here we just delete the block in the bucket, right? Then the bucket still exists, and if we call `get_file_cache` function now, this empty bucket may still be returned?